### PR TITLE
feat(qwen3.5): linear attention scan

### PR DESCRIPTION
**Pull request by co-esquie**
# Linear Attention Scan for Qwen3.5 — Massive Performance Boost

## References

1. **HuggingFace** — `torch_chunk_gated_delta_rule` in `transformers/models/qwen3_5/modeling_qwen3_5.py`
   - Reference canonical implementation of chunked Woodbury (WY) algorithm
   - Complete intra-chunk forward substitution solve `(I+A)^{-1}` and inter-chunk sequential state propagation
   
2. **FLA (Flash Linear Attention)** — `fla-org/flash-linear-attention` repository
   - Optimized GPU implementation with fused Triton kernels
   - `fla/ops/gated_delta_rule/chunk.py`: `chunk_gated_delta_rule_fwd_kkt_solve_kernel`, `chunk_gated_delta_rule_fwd_h`
   - Same algorithm with GPU-optimized kernel fusion
   
### Action!

https://github.com/user-attachments/assets/3054ef6f-eb72-432c-942b-9f80c0be64cb

## Summary

Implements the chunked gated delta rule (linear attention) for Qwen3.5 using pure Candle operations. This reduces the attention complexity from **O(T²)** to **O(T/chunk_size)** sequential steps with parallel chunk processing.

**Substantial improvements across the board:**
- **TTFT (Time To First Token)**: Dramatically faster prompt processing with linear complexity
- **Prefill**: 5-10x speedup on long-context prefill operations
- **Decode**: Near-constant time per token (linear vs quadratic scaling)

We should be close to llama.cpp in benchmarks for both prefill and decode throughput.

## Implementation Details

- **Chunk-based GatedDeltaNet parallel scan** (`qwen3_5_linear_attn_scan.rs`)
  - Replaces sequential O(T) recurrence with HuggingFace-proven chunked Woodbury (WY) approach
  - Chunk size: 64 tokens (configurable)
  - Handles padding and shape transformations for batch processing
  
- **Reference implementation** from HuggingFace transformers (`torch_chunk_gated_delta_rule`)
  - Proven correctness in production Qwen3.5 models
  - We replicate the exact mathematical steps in Candle

- **Pure Candle kernels** — no custom CUDA kernels required
  - Uses existing `broadcast_matmul`, `cumsum`, and tensor ops
  - Portable across CPU/GPU backends

## Questions Regarding Kernel Mathematical Accuracies

**To testers:** The implementation uses floating-point operations for:
- Log-space decay calculations (`log_g`, `g_cumsum`, decay mask exponentials)
- Forward substitution solve for `(I + A)^{-1}`
- Weighted key-value combinations

Please test on **CPU/non-CUDA backends** to verify:
1. Numerical stability of decay mask computation (exp of large negatives)
2. Accuracy of forward substitution vs. reference implementation
3. Edge cases: very small/large chunk sizes, gradient flow, inference vs. training
4. Correctness across different tensor backends

Run existing tests and compare outputs against the reference HuggingFace model.

### CUDA 5D Matmul Bug in candle (gemm_config stride_b)

**Problem:** Calling `broadcast_matmul` on a 5D tensor with 3 batch-prefix dimensions (e.g., `[b, n_heads, num_chunks, S, d]`) silently falls through to `MatMulNonContiguous`, causing a runtime error.

**Root cause:** In `candle-core/src/cuda_backend/mod.rs`, the `gemm_config` function matches on the length of the batch-stride slice:
```rust
match batch_strides {
    [] => /* 2D matmul */,
    [stride] => /* 1 batch dim */,
    [_, stride] | [stride] => /* up to 2 batch dims */,
    _ => return Err(CudaError::MatMulNonContiguous),
}
```

A 5D tensor produces a 3-element batch-stride slice, triggering the `_ =>` arm and causing the error.

**Solution:** Reshape 5D matmul operations to 3D by collapsing the 3 batch dimensions:
```rust
let bhnc = b * n_heads * num_chunks;
result = lhs.reshape((bhnc, S, d))?
            .broadcast_matmul(&rhs.reshape((bhnc, d, S))?)?
            .reshape((b, n_heads, num_chunks, S, S))?;
```

This reduces the batch-stride slice to 1 element, matching the `[stride] =>` arm.

**Applied in:** All 5D matmuls in `qwen3_5_linear_attn_scan.rs`:
- `kk`: k_beta @ k^T
- `contrib`: intra-chunk attention contributions
- `value_new`: final output computation
- `w`: state update computation

### CODE REVIEW

**Comment by co-esquie** [comment]
@ericcurtin This sh*t is the most complex thing I ever did. Can you please validate the tests on your side (I believe you are on metal?) and if you can give me a benchmark result. Kernel issues were really painful to debug on my side. This PR should stay in draft until we are confident on metal

**Comment by ericcurtin** [comment]
> @ericcurtin This sh*t is the most complex thing I ever did. Can you please validate the tests on your side (I believe you are on metal?) and if you can give me a benchmark result. Kernel issues were really painful to debug on my side. This PR should stay in draft until we are confident on metal

You are the first one to look at this properly. There was some qwen 3.5 implementation here based on candle, was always broken.

This is complex for every model, getting it working, optimized working on various hardware, it's a challenge.

My advice, take this out of draft, lets try and get it reviewed and merged. Having it working on one piece of hardware is better than having it working on none.

**Comment by co-esquie** [comment]
@ericcurtin Commit https://github.com/ericcurtin/inferrs/pull/172/changes/fb7d8f3799c0c188ff2ee67a986df066886b95f6 fixes **VERY** nasty cuda bugs. I consider this implementation as accurate on both CPU & CUDA kernels. I added a shit ton of tests to debug the maths on other kernels to debug them against CPU kernel. This also fix a bug where second round message were broken due to layer corruption caused by CUDA issues

```bash
cargo run --release --bin inferrs -- run Qwen/Qwen3.5-2B --device=cuda

> Hello!
Hello! How can I help you today?
> describe docker company in one sentence
Docker is a company that creates a lightweight and flexible platform for developing, shipping, and running applications in containers.
```

```bash
cargo run --release --bin inferrs -- run Qwen/Qwen3.5-2B --device=cpu

> Hello!
Hello! How can I help you today?
> Describe docker company in 3 words
Cloud-native DevOps.
```

Also bot feedbacks are good but not exhaustive. I've added on top of the file https://github.com/ericcurtin/inferrs/pull/172/changes/fb7d8f3799c0c188ff2ee67a986df066886b95f6#diff-5df2c47969ff6bf21a6031c232d118fbabea4f52029bfaf8007cc8c41ac3bdcaR9 the remaining todos according to me. They need to be heavily unit tested to ensure maths correctness but in the current state this PR already brings a massive performance upgrade

Confidence 5/5 first try greptile 😎

EDIT: let me know what you think about tests in inferrs/src/models/qwen3_5_linear_attn_scan.rs for CI runners

And also here is the numbers:

### Before this PR

| Métrique      | inferrs     | llama.cpp   |
|---------------|-------------|-------------|
| TTFT          | 939 ms      | 51 ms       |
| Prefill       | 320 tok/s   | 5783 tok/s  |
| Decode        | 17 tok/s    | 48 tok/s    |
| Peak memory    | 630 MB      | 4079 MB     |


### After this PR

```
═════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════
Backend                                                             TTH (ms)     TTFT (ms)   Prefill (t/s)   Decode (t/s)   Peak mem (MB)
─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
llama-server -hf Mungert/Qwen3.5-0.8B-GGUF                        8015.86 ms      53.21 ms     5611.91 t/s      46.19 t/s      4076.38 MB
inferrs serve --quantize Qwen/Qwen3.5-0.8B                        6023.33 ms     216.34 ms     1391.53 t/s      39.32 t/s       748.43 MB
inferrs serve --turbo-quant=false --quantize Qwen/Qwen3.5-0.8B    6048.38 ms     216.49 ms     1390.92 t/s      44.04 t/s       744.18 MB
inferrs serve Qwen/Qwen3.5-0.8B                                    26036.32ms     222.77 ms     1352.56 t/s      38.05 t/s       782.89 MB
═════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════
```

Despite not being marketing-friendly, qwen 3.5 should be fairly usable (first message is very slow I need to investigate why)

image

File: inferrs/src/models/qwen3_5_linear_attn_scan.rs [hunk 1]
```diff
@@ -0,0 +1,404 @@
...
+            .exp()?
+            .unsqueeze(candle_core::D::Minus1)?; // [b, n_h, S, 1]
+
+        // state = state * exp(g_cumsum[-1]) + (k * decay_to_end)^T @ v_corrected
+        let k_weighted = k_ci.broadcast_mul(&decay_to_end)?; // [b, n_h, S, hk]
+        let state_update = k_weighted
+            .transpose(candle_core::D::Minus1, candle_core::D::Minus2)?
+            .broadcast_matmul(&v_corrected)?; // [b, n_h, hk, hv]
+        s = (s.broadcast_mul(&g_end_exp)? + state_update)?;
+    }
```
[the remark this answers is not in the document]
Comment (co-esquie) [reply inferrs/src/models/qwen3_5_linear_attn_scan.rs]: added as todo on top of the file with more performances updates incoming

**Comment by ericcurtin** [comment]
Can optimize in follow-on prs
